binding.cc

```diff
@@ -186,6 +186,127 @@ NAN_METHOD(crypto_kx_server_session_keys) {
   CALL_SODIUM(crypto_kx_server_session_keys(CDATA(rx), CDATA(tx), CDATA(server_pk), CDATA(server_sk), CDATA(client_pk)))
 }
 
+// crypto_aead
+
+NAN_METHOD(crypto_aead_xchacha20poly1305_ietf_keygen) {
+  ASSERT_BUFFER_MIN_LENGTH(info[0], k, crypto_aead_xchacha20poly1305_ietf_keybytes())
+  crypto_aead_xchacha20poly1305_ietf_keygen(CDATA(k));
+}
+
+// (ciphertext_buf, plaintext_buf, [ad], null, npub_buf, k_buf)
+NAN_METHOD(crypto_aead_xchacha20poly1305_ietf_encrypt) {
+  ASSERT_BUFFER_SET_LENGTH(info[1], plaintext)
+  ASSERT_BUFFER_MIN_LENGTH(info[0], ciphertext, plaintext_length + crypto_aead_xchacha20poly1305_ietf_abytes())
+  ASSERT_BUFFER_MIN_LENGTH(info[4], npub, crypto_aead_xchacha20poly1305_ietf_npubbytes())
+  ASSERT_BUFFER_MIN_LENGTH(info[5], k, crypto_aead_xchacha20poly1305_ietf_keybytes())
+
+  const unsigned char *ad_data = NULL;
+  size_t ad_len = 0;
+  if (info[2]->IsObject()) {
+    ASSERT_BUFFER_SET_LENGTH(info[2], ad)
+    ad_data = CDATA(ad);
+    ad_len = ad_length;
+  }
+
+  const unsigned char *nsec = NULL;
+  unsigned long long clen;
+  CALL_SODIUM(crypto_aead_xchacha20poly1305_ietf_encrypt(
+    CDATA(ciphertext), &clen,
+    CDATA(plaintext), plaintext_length,
+    ad_data, ad_len,
+    nsec,
+    CDATA(npub),
+    CDATA(k)
+  ))
+
+  info.GetReturnValue().Set(Nan::New((uint32_t) clen));
+}
+
+// (plaintext, null, ciphertext, [ad], npub, k)
+NAN_METHOD(crypto_aead_xchacha20poly1305_ietf_decrypt) {
```

Comment:
@emilbayes what is `info[1]`?

Reply:
A `nsec` in libsodium, which is unused for this particular cipher. There's on official `crypto_aead` primitive yet, so I guess frank made all the algo specific interfaces the same to keep options open. I didn't want to hide this detail to keep parity with the libsodium docs